File: python/open_guardrail/guards/prompt_echo.py
"""Detect when AI echoes back the prompt."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_SYSTEM_PROMPT: list[re.Pattern[str]] = [
    re.compile(
        r"\byou\s+are\s+(?:a|an)"
        r"\s+(?:helpful|friendly|professional)"
        r"\s+(?:AI|assistant|chatbot)\b",
        re.IGNORECASE,
    ),
    re.compile(r"\bsystem\s*:\s*", re.IGNORECASE),
    re.compile(
        r"\b(?:instructions?|rules?)\s*:\s*\n",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bdo\s+not\s+(?:reveal|share|disclose)"
        r"\s+(?:these?\s+)?"
        r"(?:instructions?|rules?|prompt)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\byour\s+(?:role|purpose|task)"
        r"\s+is\s+to\b",
        re.IGNORECASE,
    ),
]

_ECHO_INDICATORS: list[re.Pattern[str]] = [
    re.compile(
        r"\b(?:here\s+(?:is|are)\s+)?"
        r"(?:my|the)\s+(?:system\s+)?"
        r"(?:prompt|instructions?)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\b(?:as\s+)?instructed\s+in"
        r"\s+(?:my|the)\s+(?:system\s+)?prompt\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bmy\s+(?:original|initial)"
        r"\s+(?:instructions?|programming|prompt)\b",
        re.IGNORECASE,
    ),
]

_USER_ECHO: list[re.Pattern[str]] = [
    re.compile(
        r"^user\s*:\s*", re.IGNORECASE | re.MULTILINE
    ),
    re.compile(
        r"^human\s*:\s*",
        re.IGNORECASE | re.MULTILINE,
    ),
    re.compile(
        r"\b(?:you\s+(?:asked|said|wrote))"
        r"\s*:\s*[\"']",
        re.IGNORECASE,
    ),
]


class _PromptEcho:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "prompt-echo"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        if any(p.search(text) for p in _SYSTEM_PROMPT):
            issues.append("system_prompt_exposure")
        if any(p.search(text) for p in _ECHO_INDICATORS):
            issues.append("instruction_echo")
        if any(p.search(text) for p in _USER_ECHO):
            issues.append("user_prompt_echo")

        triggered = len(issues) > 0
        score = (
            min(len(issues) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="prompt-echo",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Prompt echo detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def prompt_echo(
    *, action: str = "block"
) -> _PromptEcho:
    return _PromptEcho(action=action)

File: packages/guards/src/prompt-echo.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface PromptEchoOptions {
  action: 'block' | 'warn';
}

const SYSTEM_PROMPT_PATTERNS: RegExp[] = [
  /\byou\s+are\s+(?:a|an)\s+(?:helpful|friendly|professional)\s+(?:AI|assistant|chatbot)\b/i,
  /\bsystem\s*:\s*/i,
  /\b(?:instructions?|rules?)\s*:\s*\n/i,
  /\bdo\s+not\s+(?:reveal|share|disclose)\s+(?:these?\s+)?(?:instructions?|rules?|prompt)\b/i,
  /\byour\s+(?:role|purpose|task)\s+is\s+to\b/i,
];

const ECHO_INDICATORS: RegExp[] = [
  /\b(?:here\s+(?:is|are)\s+)?(?:my|the)\s+(?:system\s+)?(?:prompt|instructions?)\b/i,
  /\b(?:as\s+)?instructed\s+in\s+(?:my|the)\s+(?:system\s+)?prompt\b/i,
  /\bmy\s+(?:original|initial)\s+(?:instructions?|programming|prompt)\b/i,
];

const USER_ECHO_PATTERNS: RegExp[] = [
  /^user\s*:\s*/im,
  /^human\s*:\s*/im,
  /\b(?:you\s+(?:asked|said|wrote))\s*:\s*["']/i,
];

export function promptEcho(options: PromptEchoOptions): Guard {
  return {
    name: 'prompt-echo',
    version: '0.1.0',
    description: 'Detects when AI echoes back the prompt verbatim',
    category: 'ai',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      if (SYSTEM_PROMPT_PATTERNS.some((p) => p.test(text))) {
        issues.push('system_prompt_exposure');
      }
      if (ECHO_INDICATORS.some((p) => p.test(text))) {
        issues.push('instruction_echo');
      }
      if (USER_ECHO_PATTERNS.some((p) => p.test(text))) {
        issues.push('user_prompt_echo');
      }

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / 3, 1.0) : 0;

      return {
        guardName: 'prompt-echo',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
